### DISCUSSION TITLE
Workflow Registry marked as internal

### DISCUSSION THREAD

**Discussion by sgehrig** [symfony/symfony#49199] [Q&A] [answered]
With #46000 the workflow registry has been marked as `internal` (Symfony 6.2). Unfortunately I did not follow this discussion closely, so I might have missed some things. How does this work together with being able to define `\Symfony\Component\Workflow\SupportStrategy\WorkflowSupportStrategyInterface`s per workflow? 

With the registry I was able to just ask the registry to give me a workflow that supports a given object. Something like 

```php
$workflow = $this->workflowRegistry->get($job);
```

What is the intended approach for such a use case without using the registry? I know that I can inject the workflows as either iterators or locators into any service, but how do I find the workflow supporting a given object? I'd need the accompanying support strategies for the workflows. 

I mean even the documentation (on https://symfony.com/doc/current/components/workflow.html) describes this exact scenario - **using the registry**.

```php
$blogPostWorkflow = ...;
$newsletterWorkflow = ...;

$registry = new Registry();
$registry->addWorkflow($blogPostWorkflow, new InstanceOfSupportStrategy(BlogPost::class));
$registry->addWorkflow($newsletterWorkflow, new InstanceOfSupportStrategy(Newsletter::class));

$blogPost = new BlogPost();
$workflow = $registry->get($blogPost);
```

**Comment by lyrixx** [comment, marked as answer]
Hello. Actually the registry was a bad idea. It's useful only in twig.
It's easier to inject the workflow directly where you need it.

What is you use case for using a registry and not the right workflow directly? 

Note: I submitted a PR to remove this part of the documentation: https://github.com/symfony/symfony-docs/pull/17878

**Reply by sgehrig** [reply]
Hi @lyrixx,

I'd not say, that the registry was per se a bad idea. Perhaps the name was a bit misleading, I'd say (registry sounds like just a service locator). The registry allows you to retrieve a workflow not just by its name but rather by an object (e.g. an entity) checking the defined support strategy and returning the workflow that supports the given object. The registry therefore does not only store the workflows it also stores the configured support strategy for each workflow. 

Let's say we have two different Job objects. Each one uses a different workflow - depending on some properties. Therefore we have two strategies (aka `WorkflowSupportStrategyInterface`) that select the correct workflow based on the object. The code is rather simple:

```php
$workflow = $this->registry->get($job);
```

How would I do that without accessing the registry? What's the purpose of these support strategies if we internalise the registry?

 > What is you use case for using a registry and not the right workflow directly?

In our case that'd mean to inject both workflows (or a `tagged_iterator` / `tagged_locator`) wherever we currently need the registry and build the workflow selection ourselves. Then the question would be how to access the configured support strategies?

**Reply by lyrixx** [reply]
I'm sorry but I did not understand your use case. 

Why don't you inject the right workflow directly where you need it? I mean, there is no need for the registry in most of use case

**Reply by sgehrig** [reply]
Perhaps I miss something about the changes in the workflow component and the ideas and concepts that went into them. As I said, I was not following the workflow component closely. 

Let's phrase it differently: how would you use the workflow support strategy concept without the registry?

**Reply by lyrixx** [reply]
I won't because I don't need it. I inject the right workflow where I need it directly.

For example for things that need the workflow about rule:
```php
    public function __construct(
        #[Target('rulesStateMachine')]
        private readonly WorkflowInterface $workflow,
```
etc

**Reply by sgehrig** [reply]
Ok. Given your example now imagine you have two different workflows (or X workflows) and which one to use is only defined at runtime based on the subject type or its properties. 

That's exactly the use case I imagined the workflow strategies to support. 

Without the registry (or a similar service, aka **workflow selector**) the support strategies become obsolete and workflow selection has to go into user land code. 

Unless I miss something crucial obviously.

**Reply by lyrixx** [reply]
But ATM, I really fail to see what is the real use case for something that deal with a subject and its workflow without knowing what kind of object it deal with. You are telling me you have something super generic that is able to deal with any kind of subject? In every case I use the workflow, it is used by a class that only deal with subject of type XXX. And another class deals with object of type YYYY. Thus I inject the right workflow directly.

The only place where I can imagine something super generic is for documentation or debugging. And in this situation, I think a ServiceLocator perfectly fits.

Note: If you have a real use case to share, I'll be easier to discuss. For now, everything is very theoretical...

**Reply by sgehrig** [reply]
It's not fully flexible and workflows are indeed somehow compatible or constraints are enforced on the domain model. So freely switching between workflows is somehow constrained in the domain model. 

The thing that's really fully flexible is a visualization component we use to visualize the currently active workflow and it's state on a given subject. 

Is this use case so out of this world? That's how I understood the workflow support strategies: select the correct workflow during runtime based on arbitrary conditions.

**Reply by lyrixx** [reply]
The name of the workflow is not enough? Please share some code (the service that need the registry, the strategy), so we can discuss and move forward.

**Reply by sgehrig** [reply]
Sure. It would surely be possible to change the code so that the workflow name could be sufficient. But I think it'd be unnecessary given that we already have this concept of workflow support strategies. 

I'm just a bit worried that the whole idea of internalising the registry obstructs the usage of other workflow component features.

**Reply by lyrixx** [reply]
Simpler code is always better :) that's why we "remove" some features

**Reply by sgehrig** [reply]
Absolutely. But does this imply that you intend to remove the workflow support strategies as well? Because, unless I miss something, they will not be of any use any more. 

I'm not saying that it's no possible to move our use case into user land code, I just want to understand the reasoning and the plans to come.

**Reply by lyrixx** [reply]
> Absolutely. But does this imply that you intend to remove the workflow support strategies as well? Because, unless I miss something, they will not be of any use any more.

yes, the whole part of this component is now internal. There is no need for it. As you said it yourself, you can retrieve what you want with its name. So why do you want to use something more complex?

**Reply by sgehrig** [reply]
I do not insist to use the component the way we currently use it. Perhaps I could not make that clear.

It was really only about discussing the use case and understanding how you plan to move forward. 

So thanks a lot for your time. This is much appreciated. I'll continue to watch the development on this component more closely to see how we move forward with our use of the workflow component.

**Comment by ivannemets-sravniru** [comment]
Hi! My team and I have similar concern as @sgehrig regarding Registry being marked `@internal`.

We are using `Registry` to resolve workflow in runtime for each individual instance, and for that we created a custom implementation of `WorkflowSupportStrategyInterface`.

I'll provide a simplified explanation of our use case, so you could better understand why we would need that :)

Lets imagine a service that aggregates some financial products (insurance, bank loans, etc..) provided by some financial organizations. This service implements an integration with each partner (financial organizations). Each partner has it's own API and it's own process how exactly (step by step) a customer (end user) can get a bank loan from that partner.

So, lets say in that service we have an entity called `Deal`, a simplified implementation below:
```php

#[ORM\Entity(repositoryClass: DealRepository::class)]
class Deal
{
    #[ORM\Id]
    #[ORM\Column(type: 'ulid', unique: true)]
    private Ulid $id;

    #[ORM\ManyToOne(fetch: 'EAGER')]
    #[ORM\JoinColumn(nullable: false)]
    private Integration $integration;

    #[ORM\Column(type: 'json')]
    private array $states = [];

    public function getId(): Ulid
    {
        return $this->id;
    }

    public function getIntegration(): Integration
    {
        return $this->integration;
    }

    public function setIntegration(Integration $integration): self
    {
        $this->integration = $integration;

        return $this;
    }

    public function getStates(): array
    {
        return $this->states;
    }

    public function setStates(array $states): self
    {
        $this->states = $states;

        return $this;
    }
}

```
Methods `getStates()` and `setStates()` are the methods used by `MarkingStore` to read/write object's state. The Deal can be at multiple places (states) at a time.

`Deal` has a relation to `Integration` entity which contains some information about the integration with a specific partner. An integration has a unique name, and a unique **integration code**.

The integration process (workflow) with all partners is different in general, but there are some common operations that appear in all integrations (create a Deal entity within our service, send initial user data to partner, upload user documents, etc.)

There are API endpoints for all these common operations that define all possible interactions with our customer (end user). In these API endpoints (besides other business logic which is the same for all integrations) we, eventually, resolve Deal workflow in runtime and apply a common transition (which exists in each integration's workflow). Deal workflow depends on what product from what partner the customer has selected.

```php
// somewhere within API endpoint's logic (exceptions handling omitted for simplicity)
$dealWorkflow = $this->workflowRegistry->get($deal);
$dealWorkflow->apply($deal, 'InitDeal');
```

There are few key points to take into account:
1. To resolve workflow we have a custom implementation of `\Symfony\Component\Workflow\SupportStrategy\WorkflowSupportStrategyInterface` that compares **integration_code defined in workflow metadata** and **integration code** of the given `Deal` entity (`$deal->getIntegration()->getCode()`).
```php
final class IntegrationCodeSupportStrategy implements WorkflowSupportStrategyInterface
{
    public function supports(WorkflowInterface $workflow, object $subject): bool
    {
        return $subject instanceof Deal
            && $workflow->getMetadataStore()->getMetadata('integration_code') === $subject->getIntegration()->getCode();
    }
}
```
2. A transition named `InitDeal` is defined in each workflow, but the definition is different (`to` place(s) might be different)
3. The business logic around `InitDeal` transition is different for each integration, and that logic is implemented via event subscribers for each specific workflow.

So, basically, we have multiple workflows that look similar to this:
```yaml
framework:
    workflows:
        some_partner_deal:
            type: 'workflow'
            support_strategy: 'app.workflow.integration_code_supports_strategy' # custom supports strategy
            metadata:
                integration_code: 'some_partner_integration_code' # this code is unique for all integrations
            initial_marking: 'new'
            places:
                'new': ~
                'foo': ~
                'bar': ~
                # other places...
            transitions:
                'InitDeal': # this transition gets applied within an API endpoint reused in all integrations with all partners 
                    from: 'new'
                    to: 'foo'
                'SomePartnerSpecificTransition': # this transition gets applied somewhere in partner-specific integration code
                    from: 'foo'
                    to: 'bar'
                # other transitions...
```

The real usage is more complex but I hope this should be enough to explain why we use `\Symfony\Component\Workflow\Registry`.

I'm sure we are not the only team who finds it useful, at least since what @sgehrig described in this discussion sounds like he has something similar. And there are many devs still using Symfony 5.4 and have no clue what changes are coming in 6.2+ - they might also share some interesting use cases...

@lyrixx You proposed using service locators instead of `Registry`.

I thought it over, so, yes, it's possible with some restrictions, and I guess it would look similar to this:

```php
class WorkflowResolver
{
    private ContainerInterface $workflowLocator;

    public function __construct(#[TaggedLocator('workflow')] ContainerInterface $workflowLocator) {
        $this->workflowLocator = $workflowLocator;
    }

    public function resolve(Deal $deal): WorkflowInterface
    {
        // to get a workflow instance from service locator we do need the workflow name, no other options here..
        // so I imagine we would have to use integration code as a part of workflow name
        $workflowName = sprintf('%s_deal', $deal->getIntegration()->getCode());

        // and then we just get the required workflow by its name
        // looks quite simple, right?
        return $this->workflowLocator->get($workflowName);
    }
}
```

But if you take a closer look, you may note this approach (`WorkflowResolver`) does essentially the same thing what a `WorkflowSupportStrategyInterface` is intended for - it implements the way how we define what workflow supports what Deal objects, and retrieves the workflow from service locator.

We could do the same by implementing a `WorkflowSupportStrategyInterface`:

```php
final class IntegrationCodeSupportStrategy implements WorkflowSupportStrategyInterface
{
    public function supports(WorkflowInterface $workflow, object $subject): bool
    {
        return $subject instanceof Deal
            && $workflow->getName() === sprintf('%s_deal', $subject->getIntegration()->getCode());
    }
}
```

... and still use `Registry::get()` instead of `WorkflowResolver::resolve()`.

Basically `WorkflowResolver` is a limited version of the `Registry`. `Registry` in combination with `WorkflowSupportStrategyInterface` is more flexible and allows you to implement more complex support strategies.

So, I see 2 disadvantages comparing to existing functionality:
1. Using service locators the only option we have is to come up with some specific rules when giving **names** to our workflows.
2. The approach with service locators should be implemented as a custom solution in specific projects (thought before it was a part of the component out of the box)

**The main question is simple: Why?** What is the reason to mark `Registry` as `@internal`?

As @sgehrig already mentioned, having `WorkflowSupportStrategyInterface` makes no sense without `Registry`, so this should also be marked `@internal`? And what's next? Deprecated and eventually removed? Just for simplicity?

In my experience I didn't have any issues with `Registry` and `WorkflowSupportStrategyInterface` approach. I find this concept useful (I mean the support strategy itself assigned to a specific Workflow).

Some extra thoughts: if the Core Team wants to get rid of `Registry` for some reason, maybe we could keep the idea of supports strategies by making the `WorkflowSupportStrategyInterface` to be a dependency for the `Workflow` itself?
Within `Workflow` we could call the support strategy to know if the given object is actually supported by the workflow.
We could also expose a public method like `Workflow::supports(object $subject)` that could be used in combination with tagged iterator to resolve workflow for objects in runtime and still have the flexibility of `WorkflowSupportStrategyInterface`?

The strategy could be optional, by default workflow would support all objects (in fact as it is now, devs are free to pass any object to a workflow instance, which is a good place for potential human mistakes.. so maybe it would be even better to have some built-in object check within Workflow, I guess a default check like `instanceof` would not result in notable performance decrease).

Not sure if it makes any sense for anyone, but thanks for taking time to read all this! :)

**Reply by ivannemets-sravniru** [reply]
@lyrixx Hi!
Do you think you could take a look at my comment and share any thoughts?
Thank you!

**Reply by lyrixx** [reply]
(yes, replying to you is in my todo list - big comments imply big responses, I guess :) )

**Reply by lyrixx** [reply]
First, Thanks for this very details comments. It's very helpful to understand what's going on! I have concret elements and not hypothetical theories :) 

Next, I'm sorry for the long delay!

---

Then, I'm gonna reply to your questions:

> **The main question is simple: Why?** What is the reason to mark `Registry` as `@internal`?

I decided to make the Registry *internal* because I saw a huge number of projects, with this kind of code:

```php
public function fooIndex(Workflow\Registry $r, Article $a)
{
    $r->get($a)->apply('foo');
}
```

This does not make sens to me because:

1. It does not respect the [law of Demeter](https://en.wikipedia.org/wiki/Law_of_Demeter) because it adds an extra layer
1. It's harder to tests
1. It instantiates all workflows to get the right one. So it breaks lazy loading and decrease performance.
1. It make code harder to understand (more code == more bug)


Even the Symfony documentation was wrong at some point about it ; and it made me sad (I fixed it)

So I made it internal to avoid this kind of stuff. To me it's much easier to get the workflow you need instead of a black box. 

More over, other parts of the framework don't provide such *Registry* too. For exemple, there is no Registry for the Form, HttpClient, Lock, Semaphore, etc. components.

---

That's being said, let's take a close look at your application.

I understand you have different workflows. And you can know, only at runtime, which one you must use. And indeed the Registry could be helpful here!

But, the `WorkflowResolver` you wrote is so much simpler and performant than the Registry. There is less code, and you have built-in lazy loading. It scales much better!

I, really think people should write their own implementation of the Registry / Resolver / What Ever that fits their need; than a too generic solution.

Note: `WorkflowSupportStrategyInterface` becomes internal too :) 

---

Finally, I really hope you understand the move here. I understand this is really disappointing when you built some code on top of a library / framework, and then someone decide to remove the feature you used. It's frustrating, and unfair! I apologize for that.

But in this very situation, the impact is really small. With less than 10 LOC you implemented the same feature, but simpler and more performant! 

---

Of course, my mind is not closed and might evolves if some new arguments are raised.

**Reply by ivannemets-sravniru** [reply]
@lyrixx 
I agree, it's absolutely reasonable to get rid of `Registry`. I was more concerned about the fate of `WorkflowSupportStrategyInterface` :)

Anyway, it's not a big deal, in our project we will likely end up implementing something similar to `WorkflowResolver` I provided above.
So, other teams should be able to figure out their own solution for their use cases.

Thank you!

**Comment by broncha** [comment]
Hi @lyrixx 

I just upgraded my project and this change broke it.
Here's how I am using the registry with API Platform (very generic implementation)

a) I have a compiler pass, that looks for the `addWorkflow` method call in registry and prepares a list of entities that have a workflow

b) Then I have a factory that creates 2 new endpoints for these entities at runtime. 

- 1 to get the enabled transitions (GET operation)
- 2 to apply transition (PATCH operation)

(I would have no idea about the workflow names here!)

c) Then I have listeners who listen to these operations and perform the transitions. 

(I would have no idea about the workflow names here too)

For the step c I could use a service locator as you have described above, but how do you suggest I go about steps (a) and (b)

**Reply by lyrixx** [reply]
Every workflow is tagged, so you can get them from a compiler pass. It should solves your issue.

**Reply by broncha** [reply]
Thanks for replying! I tried the tagged services as well. But I could not find a way to get the entities supported for the workflow!
Also to generate the routes that responds with the available transitions for an entity, I would still need to get the workflows enabled for a given entity.

**Reply by broncha** [reply]
@lyrixx  I don't see a workaround other than making the entities aware of workflows. Do you think you'll reconsider the registry OR implement any other ways to support generic scenarios like mine?

**Reply by lyrixx** [reply]
From the tag,you can get the service, then the definition, then the supported class. With this it should be okay ?

**Reply by broncha** [reply]
The definition only has information about places, transitions, initial places, and the metadata store. 
And for generating the routes, I still need all the workflows associated with the given entity `registry->all($object)`

I take it registry is not coming back, and I have to make the entities aware of workflows anyway.

**Reply by lyrixx** [reply]
Oh you are right! And if we expose the configuration in the tag attribute, it would solves your issue, right?

**Comment by broncha** [comment]
It would for the compiler pass.
I guess then i could build my own registry in that compiler pass, which i
can use to get workflows for an entity to generate routes ( step b as
described above)

**Comment by lyrixx** [comment]
hi all, I have decided to remove the deprecation about the registry! I don't like it, but it cause too much pain I guess!


https://github.com/symfony/symfony/pull/51976
